tensorflow/python/ops/resource_variable_ops.py

```diff
@@ -1822,6 +1824,359 @@ def _init_from_proto(self, variable_def, import_scope=None):
+            with ops.device(handle.device):
+              with ops.control_dependencies([
+                gen_resource_variable_ops.assign_variable_op(
+                  handle, self.prefetch_values,
+                  name="AssignBeforeInitRead")]):
+                value = gen_resource_variable_ops.read_variable_op(handle,
+                                                                   dtype)
+                _maybe_set_handle_data(dtype, handle, value)
+              _maybe_set_handle_data(dtype, handle, value)
+            graph_element = value
+            if caching_device is not None:
+              # Variables may be created in a tf.device() or ops.colocate_with()
+              # context. At the same time, users would expect caching device to
+              # be independent of this context, and/or would not expect the
+              # current device context to be merged with the caching device
+              # spec.  Therefore we reset the colocation stack before creating
+              # the cached value. Note that resetting the colocation stack will
+              # also reset the device stack.
+              with ops.colocate_with(None, ignore_existing=True):
+                with ops.device(caching_device):
+                  cached_value = array_ops.identity(value)
+            else:
+              cached_value = None
+        else:
+          gen_resource_variable_ops.assign_variable_op(handle, initial_value)
+          is_initialized_op = None
+          initializer_op = None
+          graph_element = None
+          if caching_device:
+            with ops.device(caching_device):
+              cached_value = gen_resource_variable_ops.read_variable_op(
+                handle, dtype)
+              _maybe_set_handle_data(dtype, handle, cached_value)
+          else:
+            cached_value = None
+        if not context.executing_eagerly():
+          # Eager variables are only added to collections if they are part of an
+          # eager variable store (otherwise in an interactive session they would
+          # hog memory and cause OOM). This is done in ops/variable_scope.py.
+          ops.add_to_collections(collections, self)
+        elif ops.GraphKeys.GLOBAL_STEP in collections:
+          ops.add_to_collections(ops.GraphKeys.GLOBAL_STEP, self)
+      initial_value = initial_value if self._in_graph_mode else None
+      super(ResourceVariable, self).__init__(
+        trainable=trainable, shape=shape, dtype=dtype, handle=handle,
+        synchronization=synchronization, constraint=constraint,
+        aggregation=aggregation, distribute_strategy=distribute_strategy,
+        name=name, unique_id=unique_id, handle_name=handle_name,
+        graph_element=graph_element, initial_value=initial_value,
+        initializer_op=initializer_op, is_initialized_op=is_initialized_op,
+        cached_value=cached_value)
+
+  def update_op(self):
+    return self.params.upsert(self.ids, self.read_value(False))
```

Comment:
If multi workers update the same embeddings, only one update will take effect? Or maybe one worker read the embeddings for training, and return grad with a long time delay, the pre-trained embeddings may be covered by this update.

Reply:
> If multi workers update the same embeddings, only one update will take effect? Or maybe one worker read the embeddings for training, and return grad with a long time delay, the pre-trained embeddings may be covered by this update?

This is a good question, that's a common problem on asynchronous training, and we do some little optimization : read again on apply gradients : [here](https://github.com/tensorflow/tensorflow/pull/41371/files#diff-1379e0a008a7ef2d3b618b6909ab76eaR198)

Reply:
One more question, how to support synchronous training?

Reply:
> And another question, does this support synchronous training?

Yes, The RFC is compatible with all distributed strategy of TensorFlow,  not only PS-Worker mode.

Reply:
How will the local variables on each replicas sync with each other and push the gradients to lookup hash table in terms of synchronous training?
Plus, as one more time of I/O has been introduced due to local variables, will it cause training speed degradation?

Reply:
> How will the local variable on each replicas sync with each other and push the gradients to lookup hash table in terms of synchronous training?
> Plus, as one more time of I/O has been introduced due to local variables, will it cause training speed degradation?

This paper maybe helpful : https://arxiv.org/abs/1909.04823